examples/passport-login/src/services/user.service.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/example-passport-login
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {repository} from '@loopback/repository';
+import {Profile as PassportProfile} from 'passport';
+import {UserIdentityService} from '@loopback/authentication';
+import {UserRepository} from '../repositories';
+import {User} from '../models';
+import {UserIdentityRepository} from '../repositories/user-identity.repository';
+
+/**
+ * User service to accept a 'passport' user profile and save it locally
+ */
+export class PassportUserIdentityService
+  implements UserIdentityService<PassportProfile, User> {
+  constructor(
+    @repository(UserRepository)
+    public userRepository: UserRepository,
+    @repository(UserIdentityRepository)
+    public userIdentityRepository: UserIdentityRepository,
+  ) {}
+
+  /**
+   * find a linked local user for an external profile
+   * create a local user if not created yet.
+   * @param email
+   * @param profile
+   * @param token
+   */
+  async findOrCreateUser(profile: PassportProfile): Promise<User> {
+    if (!profile.emails || !profile.emails.length) {
+      throw new Error('email-id is required in returned profile to login');
+    }
+    const email = profile.emails[0].value;
+
+    const users: User[] = await this.userRepository.find({
+      where: {
+        email: email,
+      },
+    });
+    let user: User;
+    if (!users || !users.length) {
+      user = await this.userRepository.create({
+        email: email,
+        name: profile.displayName,
+        username: email,
+      });
+    } else {
+      user = users[0];
+    }
+    user = await this.linkExternalProfile('' + user.id, profile);
+    return user;
+  }
+
+  /**
+   * link external profile with local user
+   * @param userId
+   * @param userIdentity
+   */
+  async linkExternalProfile(
+    userId: string,
+    userIdentity: PassportProfile,
+  ): Promise<User> {
+    let profile;
+    try {
+      profile = await this.userIdentityRepository.findById(userIdentity.id);
```

Comment:
I think in UserIdentityRepo and UserCredentialsRepo, you would need to override the findById to locate using the `userId` field (instead of id field), wouldn't you? The value of User.id is stored in UserIdentity.userId and UserCredential.userId since this is the foreignKey pointing back to user right? I think it is working right now because everything has id 1 and it works. Or perhaps you can use  .find and pass a where clause where userId = something.

Reply:
@emonddr the UserIdentity.id is the id generated and available in the external oauth profile. 
the UserCredential.id is the username , that is already available in the User model. These are ideally the identifiers of these models. The uniqueness of the models are also guaranteed.

Reply:
can you produce a diagram displaying a User model and all its field, and  the id field and userId field for the UserIdentity and UserCredential models. I am having a hard time see that the relations are going to work. thx.

Reply:
Followed the conversation here, a suggestion for `UserIdentity`: maybe rename it to something that clearly say it's an EXTERNAL identity?

Like `ExternalIdentity`, `ThirdPartyIdentity`? just to distinguish it with the local `UserPorfile`.

Same for the newly added interface `UserIdentity` in `@loopback/authentication`.
WDYT? cc @deepakrkris @emonddr 

Feel free to ignore it if it's only my own concern :) I understand that renaming takes huge refactor effort.

Reply:
> UserIdentity is a term borrowed from the original loopback-component-passport , since this is a migration story, may be better to have it for now.

Fair enough 👍